# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/remoteworkers/RemoteWorkerChild.h
# Commit: f2debe329bca
# Full Hash: f2debe329bcaaad1248a234c165ccf240d602344
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2025-04-15 21:21:06
# Regressor Bug: 1899503
# File Overlap Count: 4
# Description:
#   Bug 1899503 - P3 PRemoteWorkerDebugger binding implementation. r=asuth
#   
#   This patch implements the PRemoteWorkerDebugger binding.
#   
#   We are trying to keep the same logic debugger registration/unregistration to minimize the difference between the local debugger(the debugger on the content process main thread)
# ==============================================================================

diff -r 604c4c52574a -r f2debe329bca dom/workers/remoteworkers/RemoteWorkerChild.h
--- a/dom/workers/remoteworkers/RemoteWorkerChild.h	Tue Apr 15 09:47:39 2025 +0000
+++ b/dom/workers/remoteworkers/RemoteWorkerChild.h	Tue Apr 15 09:47:39 2025 +0000
@@ -84,6 +84,11 @@
 
   const nsTArray<uint64_t>& WindowIDs() const { return mWindowIDs; }
 
+  void SetIsThawing(const bool aIsThawing) { mIsThawing = aIsThawing; }
+  bool IsThawing() const { return mIsThawing; }
+  void PendRemoteWorkerOp(RefPtr<RemoteWorkerOp> aOp);
+  void RunAllPendingOpsOnMainThread();
+
  private:
   class InitializeWorkerRunnable;
 
@@ -173,6 +178,23 @@
   };
 
   ThreadBound<LauncherBoundData> mLauncherData;
+
+  // Thaw operation holds mState.lock. It means other operations will be blocked
+  // until mState.lock is released. However, Thaw operation is blocked by
+  // RemoteWorkerDebugger registration that needs WorkerLauncher thread to send
+  // IPC to continue the registration on the parent process. If a RemoteWorkerOp
+  // is received on WorkerLauncher thread when the RemoteWorker is thawing, a
+  // deadlock could be happen between WorkerLauncher thread and RemoteWorker's
+  // parent thread. So mIsThawing and mPendingOps are introduced to avoid the
+  // deadlock by pending the operations when RemoteWorker is thawing.
+  //
+  // Note that these could be removed once RemoteWorkerChild off-main-thread
+  // done since the RemoteWorker's parent thread will be WorkerLauncher thread.
+  // And it means when executing WorkerPrivate::Thaw on WorkerLauncher thread,
+  // it is impossible to handle the IPC callback on WorkerLauncher thread at the
+  // same time.
+  Atomic<bool> mIsThawing{false};
+  DataMutex<nsTArray<RefPtr<RemoteWorkerOp>>> mPendingOps;
 };
 
 }  // namespace mozilla::dom